File: 1. cpp/3. array/lecture 1/single number 2.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int singleNumber(vector<int>& nums) {
        unordered_map<int, int> mpp;
        
        for(int i = 0; i < nums.size(); i++) {
            mpp[nums[i]]++;                     //[4] = 0 -> [4]++  = 1
        }
        for(auto i : mpp) {
            if(i.second == 1) {
                return i.first;
            }
        }
        
        return -1;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {3, 3, 4, 1, 2, 1, 2};
    cout << "Single Number: " << sol.singleNumber(nums) << endl;
    return 0;
}

File: 2. python/3. array-list/lecture 1/single number 2.py
from typing import List
from collections import defaultdict

class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        
        mpp = defaultdict(int) # 1. (defaultdict acts like C++ map)
        
        for num in nums:
            mpp[num] += 1  # [4]= 0 -> [4]++ = 1

        for key, value in mpp.items():
            if value == 1:
                return key
        
        return -1

if __name__ == "__main__":
    sol = Solution()
    nums = [4, 1, 2, 1, 2]
    print(f"Single Number: {sol.singleNumber(nums)}")